File: Kun/leetcode/3-Longest-Substring-Without-Repeating-Characters/3-Longest-Substring-Without-Repeating-Characters.go
package main

import (
	"fmt"
	"strings"
)

func lengthOfLongestSubstring(s string) int {
	// We need to calculate the substring with non-repeating characters
	// Starting from [0, 0) to [0, j), namely substring ij, when j is already in ij
	// we increate i to the pos of last charAt(j) + 1
	// Then continue increase j
	// We can do this using map
	// Store char and pos in map, when last pos of char j is bigger than i, we update i
	// get the max j - i + 1 to the ans
	valPosMap := make(map[rune]int)
	i := 0
	ans := 0
	for j, c := range s {
		if existPos := valPosMap[c]; existPos != 0 || strings.Index(s, string(c)) == 0 {
			i = existPos + 1
		}
		if j-i+1 > ans {
			ans = j - i + 1
		}
		valPosMap[c] = j
	}
	return ans
}

func main() {
	fmt.Println(strings.Index("abcabcbb", "a"))
	fmt.Println(lengthOfLongestSubstring("abcabcbb"))
}

File: Kun/leetcode/3-Longest-Substring-Without-Repeating-Characters/3-Longest-Substring-Without-Repeating-Characters.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        // Sliding window, [i, j)
        // Starting from i = 0, j = 0
        // Increment j, adding pair[char, pos] into map
        // Check if s.charAt(j) is already in the map
        // if not, increase anwser to j - i + 1
        // if so, increase i to the i pos in map + 1
        // if returned s.charAt(j) pos from map is less than i, we should ignore it
        // of returned s.charAt(j) pos from map is bigger than i, we ack it and update its pos in map
        // then set anwser to Math.max(j - i + 1, anwser)
        Map<Character, Integer> valPosMap = new HashMap<>();
        int anwser = 0;
        for (int i = 0, j = 0; j < s.length(); j++){
            Character jChar = s.charAt(j);
            Integer existPos = valPosMap.get(jChar);
            if (existPos != null && existPos >= i) {
                i = existPos + 1;
            }
            anwser = Math.max(j - i + 1, anwser);
            valPosMap.put(jChar, j);
        }
        return anwser;
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        int res = s.lengthOfLongestSubstring("pwwkew");
        System.out.println(res);
    }
}